images/cfw/opt/x1plus/lib/python/x1plus/utils.py

```diff
@@ -16,3 +16,26 @@ def serial_number():
     return subprocess.check_output(["bbl_3dpsn"], stderr=subprocess.DEVNULL).decode(
         "utf-8"
     )
+
+
+def get_MAC() -> str:
+    """Return the MAC address of the printer."""
+    if is_emulating():
+        return "CC:BD:D3:00:3B:D5"
+    # Get the device id of the currently-active network device.
+    devices = subprocess.Popen(["ip", "link"], stdout=subprocess.PIPE)
+    # The active device _should_ be broadcasting.
+    active = subprocess.run(["grep", "BROADCAST"], stdin=devices.stdout, capture_output=True)
+    identifier = active.stdout.decode().split(":")[1].strip()
+    device = subprocess.Popen(["ip", "a", "s", identifier], stdout=subprocess.PIPE)
+    # Should be "ether" even if it's a wireless device.
+    ether = subprocess.run(["grep", "ether"], stdin=device.stdout, capture_output=True)
+    return ether.stdout.decode().strip().split(" ")[1]
```

Comment:
let me think about how to do this without invoking `grep` at least :-) it might be the case that calling open `ip a s` by hand is the best we can do but we definitely can do better than calling `grep`

Reply:
It's probably easiest if you check whether you need the address for eth0 or wlan0 ahead of time and then do
```
        with open(f'/sys/class/net/wlan0/address', 'r') as file:
            mac_address = file.read().strip()
        return mac_address
```

Reply:
Yeah, it turns out I need to use the same MAC whatever one is actually active, so I'll just pick wlan.

Reply:
> It's probably easiest if you check whether you need the address for eth0 or wlan0 ahead of time and then do
> 
> ```
>         with open(f'/sys/class/net/wlan0/address', 'r') as file:
>             mac_address = file.read().strip()
>         return mac_address
> ```

I don't know if I missed this comment and code or if it's newer than my last commit, but it's nicer than what I have, so I'll switch to this.